# PRML-MNIST-model-selection-criteria.docx

In [28]:
# Import necessary libraries
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

# Load the diabetes dataset
diabetes = datasets.load_diabetes()
X = diabetes.data
y = diabetes.target

In [29]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [3]:
# Create a linear regression model
model = LinearRegression()
# Fit the model to the training data
model.fit(X_train, y_train)
# Make predictions on the test set
y_pred = model.predict(X_test)

In [4]:
# Calculate the mean squared error and R-squared score
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)


Let's break down what the **Mean Squared Error (MSE)** and **R-squared (R²)** outputs mean in the context of your code:

### **1. Mean Squared Error (MSE):**

- **What is it?**
  - MSE measures the average of the squared differences between the actual target values (\( y_{\text{test}} \)) and the predicted values (\( y_{\text{pred}} \)) from your model.
  - In simpler terms, it quantifies how much the model's predictions deviate from the actual values. It's a way to assess the overall error in your model's predictions.
  
- **Why squared differences?**
  - Squaring the differences ensures that errors are positive, and larger errors are penalized more heavily than smaller ones.
  
- **How is it calculated?**
  - \( \text{MSE} = \frac{1}{n} \sum_{i=1}^{n} (y_{\text{test},i} - y_{\text{pred},i})^2 \)
  - Here, \( n \) is the number of samples, \( y_{\text{test},i} \) is the true value, and \( y_{\text{pred},i} \) is the predicted value for the \( i \)-th sample.

- **Interpretation in this code:**
  - A **lower MSE** means the model’s predictions are close to the actual values. However, the absolute value depends on the scale of your target variable \( y \). In your case, a lower MSE indicates better performance.
  - If your MSE is high, it means the model’s predictions are, on average, quite far from the actual target values.

### **2. R-squared (R²) Score:**

- **What is it?**
  - R² measures how well the model's predictions explain the variability of the actual target values. It is also known as the **coefficient of determination**.
  - R² represents the proportion of variance in the dependent variable (\( y \)) that is predictable from the independent variables (\( X \)).

- **How is it calculated?**
  - \( R^2 = 1 - \frac{\text{Sum of Squared Residuals}}{\text{Total Sum of Squares}} \)
  - It compares your model’s performance against a baseline model (which always predicts the mean of the target values). If your model performs better than this baseline, the R² score will be close to 1.

- **Interpretation in this code:**
  - **R² = 1** means perfect prediction.
  - **R² = 0** means the model predicts no better than simply using the mean of the target values.
  - **R² < 0** indicates that the model performs worse than the baseline (predicting the mean).

### Example Output Interpretation:

- **Mean Squared Error:**
  - If the MSE is low (e.g., 2500), the model is predicting the target values relatively accurately, with only a small average deviation.
  - A high MSE means the model is performing poorly.

- **R-squared:**
  - If the R² score is close to 1 (e.g., 0.85), it indicates that the model explains 85% of the variance in the target variable, meaning the model fits the data well.
  - If R² is closer to 0 (e.g., 0.3), the model is not doing a great job explaining the variance in the data.

In summary:
- **MSE** tells you the average squared prediction error.
- **R²** tells you how well your model explains the variance in the data.

In [7]:
# Print the results
print("Mean squared error: {:.2f}".format(mse))
print("R-squared score: {:.2f}".format(r2))

Mean squared error: 2900.19
R-squared score: 0.45


In [22]:
import math
y_train.mean()

153.73654390934846

In [24]:
math.sqrt(mse)

53.85344583676592

The values you provided give a snapshot of your model's performance. Here’s what they indicate:

1. **Mean Squared Error (MSE):** 2900.19
   - This represents the average of the squared differences between the predicted and actual values. It indicates the magnitude of the error in your model’s predictions.

2. **Root Mean Squared Error (RMSE):** 53.85
   - This is the square root of the MSE and represents the average magnitude of the error in the same units as the target variable. It helps to interpret the error in a more intuitive way. In this case, the RMSE is approximately 53.85.

3. **R-squared Score (R²):** 0.45
   - This measures how well the model explains the variance in the target variable. An R² score of 0.45 means that approximately 45% of the variance in the target variable is explained by your model. The remaining 55% is unexplained or attributed to other factors.

4. **Mean of Target Variable:** 153.74
   - The mean of your target variable provides context for the RMSE. Comparing RMSE to the mean can give you an idea of how large the error is relative to the average value of the target variable.

**Interpretation:**
- The RMSE (53.85) is a substantial proportion of the mean target value (153.74), suggesting that the errors are relatively large compared to the average target value.
- An R² score of 0.45 indicates that your model is moderately good but still has room for improvement. About 55% of the variance in the target variable is not explained by the model, which may suggest that either the model could be improved or that other variables need to be considered.

Overall, while your model provides some insight, it might benefit from further tuning, additional features, or different approaches to improve its predictive accuracy.

Let's break down the figures:

### **Mean Squared Error (MSE): 2900.19**

- **What does this mean?**
  - This value indicates the average squared difference between the predicted values and the actual values. On average, the squared errors between your model's predictions and the true target values is 2900.19.
  
- **Interpretation:**
  - The MSE being 2900.19 suggests that the model's predictions are somewhat off from the true target values, and the magnitude of the error is relatively large. However, the absolute meaning of this number depends on the scale of your target variable, which in this case is the diabetes dataset's target values (a measure of disease progression).

### **R-squared (R²) Score: 0.45**

- **What does this mean?**
  - The R² score tells you how well the model explains the variance in the target variable. A score of **0.45** means that the model explains 45% of the variance in the target values.

- **Interpretation:**
  - An R² score of **0.45** is relatively low. It means that while the model is capturing some patterns in the data, it's missing quite a lot. In fact, 55% of the variability in the data is still unexplained by the model.
  - A higher R² value would indicate a better fit (closer to 1), whereas a value close to 0 would suggest the model isn't capturing the data patterns well.

### Summary:
- **MSE** (2900.19) shows there's some significant error in your predictions, and the model isn't very accurate.
- **R² score** (0.45) suggests that while the model captures some of the data's variability, it's missing more than half of the information needed to make precise predictions. It’s a sign that the model can be improved, potentially through better feature selection, more complex models, or further tuning.

In this context, you might consider experimenting with other regression models or feature engineering to improve performance.

In [6]:
# Print the coefficients and intercept of the model
print("\nModel Coefficients:")
for feature, coef in zip(diabetes.feature_names, model.coef_):
    print(f"{feature}: {coef:.4f}")
print(f"Intercept: {model.intercept_:.4f}")


Model Coefficients:
age: 37.9040
sex: -241.9644
bmi: 542.4288
bp: 347.7038
s1: -931.4888
s2: 518.0623
s3: 163.4200
s4: 275.3179
s5: 736.1989
s6: 48.6707
Intercept: 151.3456


In [25]:
from sklearn.model_selection import cross_val_score, GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC

In [31]:
# Define models to try
models = [
    ('Logistic Regression', LogisticRegression()),
    ('Decision Tree', DecisionTreeClassifier()),
    ('Random Forest', RandomForestClassifier()),
    ('SVM', SVC())
]


In [32]:
# Compare models using cross-validation
for name, model in models:
    scores = cross_val_score(model, X_train, y_train, cv=5, scoring='accuracy')
    print(f"{name}: {scores.mean():.4f} (+/- {scores.std() * 2:.4f})")


C:\Users\James\AppData\Roaming\Python\Python311\site-packages\sklearn\model_selection\_split.py:776: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(
C:\Users\James\AppData\Roaming\Python\Python311\site-packages\sklearn\model_selection\_split.py:776: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(
C:\Users\James\AppData\Roaming\Python\Python311\site-packages\sklearn\model_selection\_split.py:776: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(


Logistic Regression: 0.0113 (+/- 0.0113)
Decision Tree: 0.0000 (+/- 0.0000)
Random Forest: 0.0057 (+/- 0.0229)


C:\Users\James\AppData\Roaming\Python\Python311\site-packages\sklearn\model_selection\_split.py:776: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(


SVM: 0.0113 (+/- 0.0211)


In [33]:
# Example of hyperparameter tuning for the best model (let's say it's Random Forest)
rf_params = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10]
}

In [35]:
rf_grid = GridSearchCV(RandomForestClassifier(), rf_params, cv=5, scoring='accuracy')
rf_grid.fit(X_train, y_train)

C:\Users\James\AppData\Roaming\Python\Python311\site-packages\sklearn\model_selection\_split.py:776: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(


GridSearchCV(cv=5, estimator=RandomForestClassifier(),
             param_grid={'max_depth': [None, 10, 20, 30],
                         'min_samples_split': [2, 5, 10],
                         'n_estimators': [100, 200, 300]},
             scoring='accuracy')

In [36]:
print("Best parameters:", rf_grid.best_params_)
print("Best cross-validation score:", rf_grid.best_score_)

# Final model
best_model = rf_grid.best_estimator_

Best parameters: {'max_depth': None, 'min_samples_split': 10, 'n_estimators': 100}
Best cross-validation score: 0.014245472837022133


Certainly! This code snippet is used for evaluating and tuning machine learning models. Let me break it down step by step:

### 1. Importing Libraries
```python
from sklearn.model_selection import cross_val_score, GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
```
- **`cross_val_score`**: Used for evaluating the performance of a model using cross-validation.
- **`GridSearchCV`**: Used for hyperparameter tuning by exhaustively searching over a specified parameter grid.
- **`LogisticRegression`, `DecisionTreeClassifier`, `RandomForestClassifier`, `SVC`**: These are different machine learning classifiers.

### 2. Defining Models
```python
models = [
    ('Logistic Regression', LogisticRegression()),
    ('Decision Tree', DecisionTreeClassifier()),
    ('Random Forest', RandomForestClassifier()),
    ('SVM', SVC())
]
```
- This list contains tuples where each tuple represents a model with its name and instance. Four models are included: Logistic Regression, Decision Tree, Random Forest, and Support Vector Machine (SVM).

### 3. Comparing Models Using Cross-Validation
```python
for name, model in models:
    scores = cross_val_score(model, X_train, y_train, cv=5, scoring='accuracy')
    print(f"{name}: {scores.mean():.4f} (+/- {scores.std() * 2:.4f})")
```
- **`cross_val_score`**: Evaluates each model using 5-fold cross-validation (`cv=5`), which means the dataset is split into 5 parts, and the model is trained on 4 parts while testing on the remaining part. This process is repeated 5 times.
- **`scoring='accuracy'`**: Measures the accuracy of the model.
- The output shows the mean accuracy score for each model and the standard deviation of these scores (multiplied by 2 to give a 95% confidence interval).

### 4. Hyperparameter Tuning for the Best Model
```python
rf_params = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10]
}
rf_grid = GridSearchCV(RandomForestClassifier(), rf_params, cv=5, scoring='accuracy')
rf_grid.fit(X_train, y_train)
print("Best parameters:", rf_grid.best_params_)
print("Best cross-validation score:", rf_grid.best_score_)
```
- **`rf_params`**: Defines the hyperparameters to be tuned for the Random Forest model, such as the number of trees (`n_estimators`), maximum depth of the trees (`max_depth`), and the minimum number of samples required to split an internal node (`min_samples_split`).
- **`GridSearchCV`**: Performs an exhaustive search over the specified hyperparameters using 5-fold cross-validation.
- **`rf_grid.fit(X_train, y_train)`**: Fits the grid search to the training data.
- **`rf_grid.best_params_`**: Outputs the best hyperparameters found during the search.
- **`rf_grid.best_score_`**: Shows the best cross-validation score achieved with the best parameters.

### 5. Final Model
```python
best_model = rf_grid.best_estimator_
```
- **`best_model`**: Stores the Random Forest model with the best-found hyperparameters. This can be used for final evaluation or predictions.

In summary, this code performs model comparison using cross-validation to evaluate the accuracy of various classifiers, then tunes hyperparameters for the best-performing model (Random Forest in this case) to improve its performance further.